### PR TITLE
Add BioImage.io Modelzoo export

### PR DESCRIPTION
**Pull request by tomburke-rse**
This is a basic API to export the keras model in addition to the current 0.3.0 BioImage.io Modelzoo spec model.
This is currently done in N2V, but should be included in CSBDeep to be available for other training schemes. 
The Modelzoo spec is in this PR matches the support in Fiji and supports the latest features there.

### CODE REVIEW

**Comment by frauzufall** [comment]
@tomburke-rse We still need to add something like `export_bioimage_io` to the default model or directly replace [`export_TF`](https://github.com/CSBDeep/CSBDeep/blob/master/csbdeep/models/care_standard.py#L201), like we did [here](https://github.com/juglab/n2v/blob/master/n2v/models/n2v_standard.py#L433-L494) for N2V. There `ModelZooBaseData` needs to be created based on the CSBDeep model (using percentile pre/postprocessing, which I still have to look up how to specify that) Then it would call the new `modelzoo_export` method.

Just a note, it will take another week before the Fiji plugins are updated to be able to fully handle the new format, so in case we replace `export_TF` we should wait until then to merge, (if @uschmidt83 is ok with this PR).

On the other hand, we could do that in a separate PR.

**Comment by uschmidt83** [comment]
Thanks for the PR (and reminding me that this still has to be done).
I hope to find some time to review this soon.

Related, note that TensorFlow model export [seems to broken](https://github.com/mpicbg-csbd/stardist/issues/90) with current stable release of TensorFlow (2.3.1).

File: csbdeep/io/modelzoo.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
+import os
+from datetime import datetime
+from hashlib import sha256
+from zipfile import ZipFile
+
+from ruamel.yaml import YAML
```
Comment (uschmidt83) [inline csbdeep/io/modelzoo.py:6]: Dependency needs to be added to setup.py

**Comment by uschmidt83** [comment]
Just seeing that the code uses type hints, which are a Python 3.5+ feature.
(As of now, csbdeep also supports Python 2.7)

**Comment by tomburke-rse** [comment]
@frauzufall I'll get to that tomorrow and look into it.
@uschmidt83 I added the only non-standard dependency and removed the type hints. Should be 2.7 compatible now.

**Comment by tomburke-rse** [comment]
@uschmidt83 There are still a few todos left. Could you help me fill them in?
Namely example images, a cover img and some values for pre- and post-processing.

I filled in everything else as good as I could, but feel free to comment if you thing something is improveable.

**Comment by uschmidt83** [comment]
Hi @tomburke-rse,

thanks for continuing to work on this PR.

> @uschmidt83 There are still a few todos left. Could you help me fill them in?
> Namely example images, a cover img and some values for pre- and post-processing.

I don't really have an overview of what's left to be done. But as far as I understand, we need to set values like "halo", etc. for all the different CARE models, which might be tricky and needs testing.

> I filled in everything else as good as I could, but feel free to comment if you thing something is improveable.

I think it would be best to keep the old `export_TF` method as is, and instead add a new `export_bioimageio` (or similar) for the new functionality.

I'll try to find some time to look over everything and it would then probably be best to have short meeting before filling in the details. What do you say?

Best,
Uwe

**Comment by tomburke-rse** [comment]
The things left to do are literally all the TODOS I left in the code :)
We can go over it whenever you have time and of course can split the export function like this, no problem at all.

File: setup.py [hunk 2]
```diff
@@ -54,6 +54,7 @@
           "tqdm",
           "pathlib2; python_version<'3'",
           "backports.tempfile; python_version<'3.4'",
+          "ruamel.yaml"
```
Comment (uschmidt83) [inline setup.py:57]: `ruamel.yaml` doesn't support Python 2.7 anymore since version [0.17.2](https://pypi.org/project/ruamel.yaml/0.17.2/). Is that version sufficient?

File: csbdeep/utils/tf.py [hunk 3]
```diff
@@ -159,8 +159,8 @@ def _export(model):
                 _model.set_weights(weights)
                 _export(_model)
 
-        if meta is not None and len(meta) > 0:
-            save_json(meta, os.path.join(dirname,'meta.json'))
+    ##    if meta is not None and len(meta) > 0:
+    ##        save_json(meta, os.path.join(dirname,'meta.json'))
```
Comment (uschmidt83) [inline csbdeep/utils/tf.py:163]:
```suggestion
        if meta is not None and len(meta) > 0:
            save_json(meta, os.path.join(dirname,'meta.json'))
```

Comment (uschmidt83) [reply csbdeep/utils/tf.py:163]: Let's not change this file.

Am I allowed to edit the code in your PR?

File: csbdeep/io/modelzoo.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        self[weight_format]["authors"] = authors
+        self[weight_format]["source"] = source
+        with open(modelfile_name, "rb") as f:
+            bytes = f.read()
+            self[weight_format]["sha256"] = sha256(bytes).hexdigest()
+        self[weight_format]["tag"] = tag
+        self[weight_format]["tensorflowversion"] = tensorflowversion
+
+
+class ModelZooBaseData(dict):
```
Comment (uschmidt83) [inline csbdeep/io/modelzoo.py:81]: Is the file copied from somewhere?

File: csbdeep/utils/tf.py (hunk 3 above)
Comment (tomburke-rse) [reply csbdeep/utils/tf.py:163]: Absolutely, please do so if necessary.

File: setup.py (hunk 2 above)
Comment (tomburke-rse) [reply setup.py:57]: I can't see why not at the moment. I'll test it tomorrow and see if something breaks

File: csbdeep/io/modelzoo.py (hunk 4 above)
Comment (tomburke-rse) [reply csbdeep/io/modelzoo.py:81]: Unfortunately not. I'll ask Wei and CC you if we already have something for all of that in place somewhere as a library.

**Comment by uschmidt83** [comment]
> The things left to do are literally all the TODOS I left in the code :)

I feel the hard work starts now, namely to automatically determine the normalization and required parameters (min_val, step_val, halo_val, etc.) and, most importantly, how to test that these parameters are actually correct (for all the different CARE models). I'd estimate that this is a few days of work.

Is there already a way to test my model after it has been exported?

File: csbdeep/io/modelzoo.py (hunk 4 above)
Comment (uschmidt83) [reply csbdeep/io/modelzoo.py:81]: No problem, I was just curious.

**Comment by uschmidt83** [comment]
I checked out the PR with Github's CLI tools, and changed the code a bit.
When I pushed the changes, they landed in the branch https://github.com/CSBDeep/CSBDeep/tree/feature/modelzoo_export. I don't know how to best work together with you on this feature. Can I directly push to https://github.com/tomburke-rse/CSBDeep/tree/feature/modelzoo_export?

**Comment by tomburke-rse** [comment]
I invited you as collaborator on the fork, you should be able to push directly on it. If that's not working, we can figure something else out.

You can currently check if your model description file is valid through this: https://github.com/bioimage-io/python-bioimage-io 
Fiji currently only supports 0.3.0, not 0.3.1. It needs an update though. 
Then you could use Fiji > Plugins >bioimage.io > bioimage.io prediction. It's one of the last things Debo did.

**Comment by maweigert** [comment]
Hi @tomburke-rse ,

thanks for the PR! 

I was wondering whether it makes sense to first adding a valid modelszoo export to stardist, before we tackle the potentially more complex integration into csbdeep itself. Advantages would be    
* We have already many trained models that we could bundle for a demo 
* It might be a more relevant use case for a modelzoo integration

I then would start a branch in stardist, trying to get a minimal modelzoo export to work, and then backport the relevant parts into this PR. Would that be ok with you?

**Comment by tomburke-rse** [comment]
Hi @maweigert 
that sounds great! You can look at this: https://github.com/bioimage-io/spec-bioimage-io/blob/move_spec/bioimageio/spec/utils/build_spec.py#L215
For a model definition writer currently in development. It'll take a few more weeks it seems until it's done, but should already be usuable. As soon as they got the a releaseable version, I'd integrate that here too and scratch the stuff I wrote last year. :)

**Comment by maweigert** [comment]
> For a model definition writer currently in development. It'll take a few more weeks it seems until it's done, but should already be usuable.

Yes, that would be wonderful. However it appears that `build_spec` doesn't support keras atm: https://github.com/bioimage-io/spec-bioimage-io/issues/87

**Comment by maweigert** [comment]
for reference, modelzoo export is now worked at here: https://github.com/stardist/stardist/pull/149

**Comment by tomburke-rse** [comment]
Since this PR is *very* stale, I'll close it now and replace it with one based on [core-bioimage-io-python ](https://github.com/bioimage-io/core-bioimage-io-python)

**Comment by uschmidt83** [comment]
I propose to first get the [StarDist integration PR](https://github.com/stardist/stardist/pull/171) done, and then think about how to transfer this to CSBDeep. Ok?
